File: fireplace-app/src/main/java/io/github/bric3/fireplace/charts/ChartComponent.java
package io.github.bric3.fireplace.charts;

import javax.swing.*;
import java.awt.*;
import java.util.Objects;

/**
 * A component that draws an {@link Chart} within the bounds of the component.
 */
public class ChartComponent extends JComponent {

    private final Chart chart;

    /** A reusable rectangle to avoid creating work for the garbage collector. */
    private final Rectangle rect = new Rectangle();

    /**
     * Creates a new component that contains an inlay chart.
     *
     * @param chart  the chart ({@code null} not permitted).
     */
    public ChartComponent(Chart chart) {
        Objects.requireNonNull(chart);
        this.chart = chart;
        setBorder(null);
    }

    /**
     * Paints the component.  The chart will be drawn at a size matching the
     * bounds of the component.
     *
     * @param g the Java2D graphics target.
     */
    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        Graphics2D g2 = (Graphics2D) g;
        getBounds(rect);
        g.translate(-rect.x, -rect.y);
        chart.draw(g2, rect);
        g.translate(rect.x, rect.y);
    }
}

File: fireplace-app/src/main/kotlin/io/github/bric3/fireplace/charts/ChartComponent.kt
/*
 * Fireplace
 *
 * Copyright (c) 2021, Today - Brice Dutheil
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package io.github.bric3.fireplace.charts

import java.awt.AWTEvent
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.Rectangle
import java.awt.event.MouseEvent
import java.beans.PropertyChangeSupport
import javax.swing.*

/**
 * A component that draws an [Chart] within the bounds of the component.
 *
 * @param chart  the chart.
 */
class ChartComponent(chart: Chart? = null) : JComponent() {
    private val propertyChangeSupport = PropertyChangeSupport(this)

    var chart: Chart? = chart
        set(value) {
            val oldChart = field
            if (oldChart == value) {
                return
            }
            field = value
            propertyChangeSupport.firePropertyChange("chart", oldChart, value)
        }

    val toolTipComponent: JComponent?
        get() = chart?.createToolTipComponent(getBounds(rect), mousePosition)

    /** A reusable rectangle to avoid creating work for the garbage collector.  */
    private val rect = Rectangle()

    init {
        enableEvents(AWTEvent.MOUSE_EVENT_MASK or AWTEvent.MOUSE_MOTION_EVENT_MASK)
        border = null
        propertyChangeSupport.addPropertyChangeListener("chart") {
            revalidate()
            repaint()
        }
    }

    /**
     * Paints the component.
     * The chart will be drawn at a size matching the bounds of the component.
     *
     * @param g the Java2D graphics target.
     */
    override fun paintComponent(g: Graphics) {
        super.paintComponent(g)
        chart?.let {
            val g2 = g as Graphics2D
            getBounds(rect)
            g.translate(-rect.x, -rect.y)
            it.draw(g2, rect, mousePosition)
            g.translate(rect.x, rect.y)
        }
    }

    override fun processMouseEvent(e: MouseEvent) {
        // handle clicks?
        super.processMouseEvent(e)
    }

    override fun processMouseMotionEvent(e: MouseEvent) {
        if (e.id == MouseEvent.MOUSE_MOVED) {
            repaint()
        }
        super.processMouseMotionEvent(e)
    }
}
